src/pybind/mgr/volumes/fs/stats_util.py

```diff
@@ -57,3 +78,252 @@ def get_stats(src_path, dst_path, fs_handle, human=True):
+
+        for clone in clones[total_clones:]:
+            with open_volume_lockless(self.volclient, clone.volname) as \
+                    fs_handle:
+                percent = get_percent_copied(clone.src_path, clone.dst_path,
+                                             fs_handle)
+                sum_percent += percent
+
+        avg_percent = round(sum_percent / total_clones, 3)
+        # progress module takes progress as a fraction between 0.0 to 1.0.
+        avg_progress_fraction = avg_percent / 100
+        msg = (f'{total_clones} ongoing+pending clones; Avg progress is '
+               f'{avg_percent}%')
+        self._update_progress_bar_event(ev_id=self.onpen_pev_id, ev_msg=msg,
+                                    ev_progress_fraction=avg_progress_fraction)
+
+        log.info('finished updating progress bar for ongong+pending clones')
+
+    def _update_ongoing_progress_bar(self, clones):
+        '''
+        Update the progress bar for ongoing cloning operations.
+        '''
+        log.info('updating progress bars for clone jobs')
+        assert self.on_pev_id is not None
+
+        percent = 0.0
+        sum_percent = 0.0
+        avg_percent = 0.0
+
+        if len(clones) < self.max_concurrent_clones:
+            total_clones = len(clones)
+        else:
+            total_clones = self.max_concurrent_clones
+
+        for clone in clones[:total_clones]:
+            with open_volume_lockless(self.volclient, clone.volname) as \
+                    fs_handle:
+                percent = get_percent_copied(clone.src_path, clone.dst_path,
+                                             fs_handle)
+                sum_percent += percent
+
+        avg_percent = round(sum_percent / total_clones, 3)
+        # progress module takes progress as a fraction between 0.0 to 1.0.
+        avg_progress_fraction = avg_percent / 100
+        msg = f'{total_clones} ongoing clones; Avg progress is {avg_percent}%'
+        self._update_progress_bar_event(ev_id=self.on_pev_id, ev_msg=msg,
+            ev_progress_fraction=avg_progress_fraction)
+
+        log.info('finished updating progress bar for ongoing clones')
+        return total_clones, sum_percent
+
+    def _update_progress_bars(self):
+        '''
+        Look for amount of progress made by all cloning operations and prints
+        progress bars, in "ceph -s" output, for average progress made
+        accordingly.
+
+        This method is supposed to be run only by instance of class RTimer
+        present in this class.
+        '''
+        clones = self._get_info_for_all_clones()
+
+        if clones:
+            total_clones, sum_percent = self._update_ongoing_progress_bar(clones)
+
+            # onpen bar (that is progress bar for clone jobs in ongoing and pending
+            # state) is printed when clones are in pending state. it is kept in
+            # printing until all clone jobs finish.
+            if len(clones) > self.max_concurrent_clones:
```

Comment:
It's a corner case, but what if `self.max_concurrent_clones` gets changed between here and L#234 (`if len(clones) < self.max_concurrent_clones:`) above? For a simplistic fix, this routine should cache the value at the start and use that throughout this routine. For a more sophisticated fix, `_get_info_for_all_clones` include the clone status and then select clones (and aggregate sizes) based on the status.

WDYT @rishabh-d-dave

Reply:
Had a conversation with Venky about this, summarizing it here. The "sophisticated fix" is to go though clone index and find states of clones. This gives us which clones are "in-progress" and which are in "pending" state. Counting number of "in-progress" clone jobs will give us the count of current number of cloner threads allowed. We'll implement this as an enhancement on a separate PR so that current code can moved towards testing and merge.

Reply:
Marking this as resolved for now.